src/fields/fe_value_handler.cc

```diff
@@ -0,0 +1,159 @@
+    value_.set_n_comp(init_data.n_comp);
+
+    if (map == nullptr) {
+		// temporary solution - these objects will be set through FieldCommon
+		map_ = new MappingP1<elemdim,3>();
+    } else {
+    	map_ = map;
+    }
+}
+
+
+template <int elemdim, int spacedim, class Value>
+typename Value::return_type const &FEValueHandler<elemdim, spacedim, Value>::value(const Point &p, const ElementAccessor<spacedim> &elm)
+{
+	ASSERT_PTR(map_).error();
+
+	Point p_rel = p - elm.element()->node[0]->point();
+	DOFHandlerBase::CellIterator cell = dh_->mesh()->element(elm.idx());
+
+	arma::mat::fixed<3,elemdim> m;
+	for (unsigned i=0; i<elemdim; ++i) {
+		m.col(i) = elm.element()->node[i+1]->point() - elm.element()->node[0]->point();
+	}
+	arma::mat::fixed<elemdim,3> im = pinv(m);
+
+	Quadrature<elemdim> quad(1);
+	quad.set_point(0, im*p_rel);
+
+	FEValues<elemdim,3> fe_values(*this->get_mapping(), quad, *dh_->fe<elemdim>(), update_values);
+	fe_values.reinit(cell);
+
+	dh_->get_loc_dof_indices(cell, dof_indices);
+
+	if (dh_->fe<elemdim>()->is_scalar()) {
+		double value = 0;
+		for (unsigned int i=0; i<dh_->fe<elemdim>()->n_dofs(); i++)
+			value += (*data_vec_)[dof_indices[i]]*fe_values.shape_value(i, 0);
+		this->value_(0,0) = value;
+	}
+	else {
+		arma::vec3 value;
+		value.zeros();
+		for (unsigned int i=0; i<dh_->fe<elemdim>()->n_dofs(); i++)
+			value += (*data_vec_)[dof_indices[i]]*fe_values.shape_vector(i, 0);
+		for (unsigned int i=0; i<3; i++)
+			this->value_(i,0) = value(i);
+	}
+
+	return this->r_value_;
+}
+
+
+template <int elemdim, int spacedim, class Value>
+void FEValueHandler<elemdim, spacedim, Value>::value_list(const std::vector< Point >  &point_list, const ElementAccessor<spacedim> &elm,
+                   std::vector<typename Value::return_type> &value_list)
+{
+	ASSERT_PTR(map_).error();
+	ASSERT_EQ( point_list.size(), value_list.size() ).error();
+
+    DOFHandlerBase::CellIterator cell = dh_->mesh()->element( elm.idx() );
+	dh_->get_loc_dof_indices(cell, dof_indices);
+
+	arma::mat::fixed<3,elemdim> m;
+	for (unsigned i=0; i<elemdim; ++i) {
+		m.col(i) = elm.element()->node[i+1]->point() - elm.element()->node[0]->point();
+	}
+	arma::mat::fixed<elemdim,3> im = pinv(m);
+
+	for (unsigned int k=0; k<point_list.size(); k++) {
+		Point p_rel = point_list[k] - elm.element()->node[0]->point();
+		Quadrature<elemdim> quad(1);
+		quad.set_point(0, im*p_rel);
+
+		FEValues<elemdim,3> fe_values(*this->get_mapping(), quad, *dh_->fe<elemdim>(), update_values);
+		fe_values.reinit(cell);
+
+		Value envelope(value_list[k]);
+
+		if (dh_->fe<elemdim>()->is_scalar()) {
+			double value = 0;
+			for (unsigned int i=0; i<dh_->fe<elemdim>()->n_dofs(); i++)
+				value += (*data_vec_)[dof_indices[i]]*fe_values.shape_value(i, 0);
+			envelope(0,0) = value;
+		}
+		else {
+			arma::vec3 value;
+			value.zeros();
+			for (unsigned int i=0; i<dh_->fe<elemdim>()->n_dofs(); i++)
+				value += (*data_vec_)[dof_indices[i]]*fe_values.shape_vector(i, 0);
```

Comment:
Make helper resolution function fe_value(fe_value, i_dof, i_qpoint), specialize for different ranks
template<dim, spacedim>
fe_value<0>(fe_value, i_dof, i_qp)
{ shape_value}

call like:
fe_value<Value::rank>(fe_values, i,0)